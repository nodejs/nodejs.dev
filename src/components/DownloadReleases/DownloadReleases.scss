.download-releases {
  width: 100%;
}

.download-releases__title {
  margin: 126px 0 50px;
  font-weight: var(--font-weight-semibold);
  font-size: var(--font-size-display4);
  line-height: var(--line-height-display4);
}

.lts__text {
  font-weight: var(--font-weight-regular);
  font-size: var(--font-size-display5);
  line-height: var(--line-height-body1);
}
