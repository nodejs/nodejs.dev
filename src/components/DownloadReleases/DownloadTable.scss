.downloads-table-container {
  overflow-x: auto;

  .downloads-table {
    color: var(--color-text-primary);
    font-size: var(--font-size-body3);

    thead {
      th {
        background: var(--color-fill-body);
        color: var(--color-text-primary);
        text-align: left;
        padding: var(--space-08) var(--space-16);
        font-weight: normal;
      }
      th:first-child {
        border-radius: 3px 0px 0px 0px;
      }
      th:last-child {
        border-radius: 0px 3px 0px 0px;
      }
    }

    tbody {
      tr {
        background: var(--color-fill-app);

        td {
          padding: var(--space-16) var(--space-16);
          border-bottom: 1px solid #e2e3e7;

          .dark-mode & {
            border-bottom: 1px solid var(--color-border-primary);
          }
        }

        &:hover {
          background: var(--brand10);
        }

        .dark-mode &:hover {
          background: var(--brand9);
        }
      }
    }
  }
}
