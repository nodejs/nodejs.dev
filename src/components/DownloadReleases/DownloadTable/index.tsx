import React from 'react';
import { FormattedMessage } from 'react-intl';
import { NodeReleaseData } from '../../../types';
import styles from './index.module.scss';

interface Props {
  nodeReleasesData: NodeReleaseData[];
}

const DownloadTable = ({ nodeReleasesData }: Props): JSX.Element => (
  <div className={styles.downloadTableContainer}>
    <table>
      <thead>
        <tr>
          <th>
            <FormattedMessage id="components.downloadTable.header.release" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.status" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.codename" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.initialRelease" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.ltsStarts" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.maintenanceStart" />
          </th>
          <th>
            <FormattedMessage id="components.downloadTable.header.endOfLife" />
          </th>
        </tr>
      </thead>
      <tbody>
        {nodeReleasesData.map(
          ({
            release,
            version,
            status,
            codename,
            initialRelease,
            ltsStart,
            maintenanceStart,
            endOfLife,
          }: NodeReleaseData): JSX.Element => {
            // Check if release is pending or not
            const isPending = status === 'Pending';
            // Download hyperlink for release
            const releaseDownloadLink = `https://nodejs.org/download/release/latest-${version}.x/`;
            // Download hyperlink for codename
            const codenameReleaseLink = `https://nodejs.org/download/release/latest-${codename.toLowerCase()}/`;

            return (
              <tr key={release}>
                <td>
                  {isPending ? (
                    version
                  ) : (
                    <a href={releaseDownloadLink}>{version}</a>
                  )}
                </td>
                <td>{status || ''}</td>
                <td>
                  {codename !== version ? (
                    <a href={codenameReleaseLink}>{codename}</a>
                  ) : null}
                </td>
                <td>{initialRelease}</td>
                <td>{ltsStart}</td>
                <td>{maintenanceStart}</td>
                <td>{endOfLife}</td>
              </tr>
            );
          }
        )}
      </tbody>
    </table>
  </div>
);

export default DownloadTable;
