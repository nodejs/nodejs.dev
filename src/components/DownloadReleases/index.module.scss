.download-releases {
  width: 100%;

  .title {
    font-size: var(--font-size-display4);
    font-weight: var(--font-weight-semibold);
    line-height: var(--line-height-display4);
    margin: 126px 0 50px;

    @media (max-width: 800px) {
      margin: 50px 0 20px 0 !important;
      text-align: center;
    }
  }
}
