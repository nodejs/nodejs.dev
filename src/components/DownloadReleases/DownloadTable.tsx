import React from 'react';

import { NodeReleaseData } from '../../hooks/useReleaseHistory';
import './DownloadTable.scss';

interface Props {
  releases: NodeReleaseData[];
}

const DownloadTable = ({ releases }: Props): JSX.Element => (
  <div className="downloads-table-container">
    <table className="downloads-table">
      <thead>
        <tr>
          <th>Release</th>
          <th>Status</th>
          <th>Codename</th>
          <th>Initial Release</th>
          <th>Active LTS Start</th>
          <th>Maintainance LTS Start</th>
          <th>End of Life</th>
        </tr>
      </thead>
      <tbody>
        {releases.map(
          ({
            release,
            status,
            codename,
            initialRelease,
            activeLTSStart,
            maintenanceLTSStart,
            endOfLife,
          }: NodeReleaseData): JSX.Element => {
            // Check if release is pending or not
            const isPending = status === 'Pending';
            // Download hyperlink for release
            const releaseDownloadLink = `https://nodejs.org/download/release/latest-${release}.x/`;
            // Download hyperlink for codename
            const codenameReleaseLink = `https://nodejs.org/download/release/latest-${codename.toLowerCase()}/`;

            return (
              <tr key={release}>
                <td>
                  {isPending ? (
                    release
                  ) : (
                    <a href={releaseDownloadLink}>{release}</a>
                  )}
                </td>
                <td>{status || ''}</td>
                <td>
                  {codename ? (
                    <a href={codenameReleaseLink}>{codename}</a>
                  ) : null}
                </td>
                <td>{initialRelease}</td>
                <td>{activeLTSStart}</td>
                <td>{maintenanceLTSStart}</td>
                <td>{endOfLife}</td>
              </tr>
            );
          }
        )}
      </tbody>
    </table>
  </div>
);

export default DownloadTable;
