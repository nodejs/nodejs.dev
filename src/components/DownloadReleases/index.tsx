import React from 'react';
import DownloadTable from './DownloadTable';
import UpcomingReleases from '../UpcomingReleases';
import { NodeReleaseData } from '../../hooks/useReleaseHistory';
import './DownloadReleases.scss';

interface Props {
  releases: NodeReleaseData[];
}

export default function DownloadReleases({ releases }: Props): JSX.Element {
  return (
    <div className="download-releases">
      <h2 className="download-releases__title">Upcoming Releases</h2>
      <UpcomingReleases />
      <p className="lts__text">
        Major Node.js versions enter Current release status for six months,
        which gives library authors time to add support for them. After six
        months, odd-numbered releases (9, 11, etc.) become unsupported, and
        even-numbered releases (10, 12, etc.) move to Active LTS status and are
        ready for general use. LTS release status is &quot;long-term
        support&quot;, which typically guarantees that critical bugs will be
        fixed for a total of 30 months. Production applications should only use
        Active LTS or Maintenance LTS releases.
      </p>
      <DownloadTable releases={releases} />
    </div>
  );
}
