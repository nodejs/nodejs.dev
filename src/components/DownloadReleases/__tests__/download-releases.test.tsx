import React from 'react';
import { render } from '@testing-library/react';
import DownloadReleases from '..';

describe('DownloadReleases component', (): void => {
  it('renders correctly', (): void => {
    const { container } = render(
      <DownloadReleases
        releases={[
          {
            release: 'v15',
            status: 'Pending',
            codename: '',
            initialRelease: '2020-10-20',
            activeLTSStart: '',
            maintenanceLTSStart: '2021-04-01',
            endOfLife: '2021-06-01',
          },
        ]}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
