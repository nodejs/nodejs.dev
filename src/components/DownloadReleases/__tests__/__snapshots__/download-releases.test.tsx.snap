// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadReleases component renders correctly 1`] = `
<div>
  <div
    class="download-releases"
  >
    <h2
      class="download-releases__title"
    >
      Upcoming Releases
    </h2>
    <div
      class="upcoming-releases"
    >
      <div
        class="react-tabs"
        data-tabs="true"
      >
        <ul
          class="react-tabs__tab-list"
          role="tablist"
        >
          <li
            aria-controls="react-tabs-1"
            aria-disabled="false"
            aria-selected="true"
            class="react-tabs__tab react-tabs__tab--selected"
            id="react-tabs-0"
            role="tab"
            tabindex="0"
          >
            Node.js 12
          </li>
          <li
            aria-controls="react-tabs-3"
            aria-disabled="false"
            aria-selected="false"
            class="react-tabs__tab"
            id="react-tabs-2"
            role="tab"
          >
            Node.js 13
          </li>
          <li
            aria-controls="react-tabs-5"
            aria-disabled="false"
            aria-selected="false"
            class="react-tabs__tab"
            id="react-tabs-4"
            role="tab"
          >
            Node.js 14
          </li>
          <li
            aria-controls="react-tabs-7"
            aria-disabled="false"
            aria-selected="false"
            class="react-tabs__tab"
            id="react-tabs-6"
            role="tab"
          >
            Node.js 15
          </li>
        </ul>
        <div
          aria-labelledby="react-tabs-0"
          class="react-tabs__tab-panel react-tabs__tab-panel--selected"
          id="react-tabs-1"
          role="tabpanel"
        >
          <div
            class="upcoming-releases__panel"
          >
            <div
              class="upcoming-releases__item--current"
            >
              <img
                alt="hexagon icon"
                src="test-file-stub"
              />
              <p
                class="release-title"
              >
                Current
              </p>
              <p
                class="release-date"
              >
                Released
                 
                2019-04-23
              </p>
            </div>
            <div
              class="upcoming-releases__item--lts"
            >
              <img
                alt="hexagon icon"
                src="test-file-stub"
              />
              <p
                class="release-title"
              >
                LTS
              </p>
              <p
                class="release-date"
              >
                Released
                 
                2019-10-21
              </p>
            </div>
            <div
              class="upcoming-releases__item--maintenance"
            >
              <img
                alt="hexagon icon"
                src="test-file-stub"
              />
              <p
                class="release-title"
              >
                Maintenance Release
              </p>
              <p
                class="release-date"
              >
                Released
                 
                2020-10-20
              </p>
            </div>
            <div
              class="upcoming-releases__item--endoflife upcoming-releases__item--to-be-released"
            >
              <img
                alt="hexagon icon"
                src="test-file-stub"
              />
              <p
                class="release-title"
              >
                End-of-life Release
              </p>
              <p
                class="release-date"
              >
                To be released
                 
                2022-04-30
              </p>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="react-tabs-2"
          class="react-tabs__tab-panel"
          id="react-tabs-3"
          role="tabpanel"
        />
        <div
          aria-labelledby="react-tabs-4"
          class="react-tabs__tab-panel"
          id="react-tabs-5"
          role="tabpanel"
        />
        <div
          aria-labelledby="react-tabs-6"
          class="react-tabs__tab-panel"
          id="react-tabs-7"
          role="tabpanel"
        />
      </div>
    </div>
    <p
      class="lts__text"
    >
      Major Node.js versions enter Current release status for six months, which gives library authors time to add support for them. After six months, odd-numbered releases (9, 11, etc.) become unsupported, and even-numbered releases (10, 12, etc.) move to Active LTS status and are ready for general use. LTS release status is "long-term support", which typically guarantees that critical bugs will be fixed for a total of 30 months. Production applications should only use Active LTS or Maintenance LTS releases.
    </p>
    <div
      class="downloads-table-container"
    >
      <table
        class="downloads-table"
      >
        <thead>
          <tr>
            <th>
              Release
            </th>
            <th>
              Status
            </th>
            <th>
              Codename
            </th>
            <th>
              Initial Release
            </th>
            <th>
              Active LTS Start
            </th>
            <th>
              Maintainance LTS Start
            </th>
            <th>
              End of Life
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              v15
            </td>
            <td>
              Pending
            </td>
            <td />
            <td>
              2020-10-20
            </td>
            <td />
            <td>
              2021-04-01
            </td>
            <td>
              2021-06-01
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
